src/lib.rs

```diff
@@ -124,6 +124,40 @@ pub fn load_native_certs() -> CertificateResult {
     }
 }
 
+/// Load certificates from the paths.
+///
+/// If both are `None`, return `Ok(None)`.
+///
+/// If `file` is `Some`, it is always used, so it must be a path to an existing,
+/// accessible file from which certificates can be loaded successfully. While parsing,
+/// the rustls-pki-types PEM parser will ignore parts of the file which are
+/// not considered part of a certificate. Certificates which are not in the right
+/// format (PEM) or are otherwise corrupted may get ignored silently.
+///
+/// If `dir` is defined, a directory must exist at this path, and all
+/// hash files contained in it must be loaded successfully,
+/// subject to the rules outlined above for `file`. The directory is not
+/// scanned recursively and may be empty.
+pub fn load_certs_from_path<T: AsRef<Path>>(file: Option<T>, dir: Option<T>) -> CertificateResult {
```

Comment:
Maybe `paths` instead of `path`? Also I would spell this as `file: Option<impl AsRef<Path>>, dir: Option<impl AsRef<Path>>` since there is no reason to require that both types are the same.

Reply:
If you use `dyn` here then the caller has to manually specify a type any time they use `None` for one of the arguments. ie `load_certs_from_path(None, Some("/tmp"))` would need to be written like `load_certs_from_path(Option::<&str>::None, Some("/tmp"))`, whereas using the same `T` generic for both means itll auto-resolve as long as at least one argument is `Some`. I figured that would be less annoying for the caller, but i agree it is more limiting and i can change this to use `dyn` if you think its best!

Reply:
actually now that i think about it more, if the caller is using two different types that satisfy `AsRef<T>`, then they can just convert them to paths on their end before passing them. which i think is less of a hassle than having to always specify `Option::<&str>::None`

Reply:
Given this, I'd suggest we just change them both to `Option<&Path>` and give up on coercion via `AsRef`. Personally I still don't like that we enforce the same type for both arguments because IMO this suggests something that really doesn't make sense (although it does work it if you just pass `&Path`, of course).